syntax = "proto3";
package exprpb;

import "metapb.proto";

enum ExprType {
    E_Invalid    = 0;

    // 列表达式
    E_Column = 1;

    // 常量值
    E_Const_Int = 2;
    E_Const_UInt = 3;
    E_Const_Double = 4;  // float or double
    E_Const_Bytes = 5;

    // 逻辑表达式
    E_LogicAnd    = 10;
    E_LogicOr     = 11;
    E_LogicNot    = 12;

    // 比较表达式
    E_Equal       = 20; 
    E_NotEqual    = 21;
    E_Less        = 22;
    E_LessOrEqual      = 23;
    E_Larger           = 24;
    E_LargerOrEqual    = 25;

    // 算数表达式
    E_Plus             = 31;
    E_Minus            = 32;
    E_Mult             = 33;
    E_Div              = 34;
}

message ColumnInfo {
    uint32 id = 1;
    metapb.DataType typ = 2;
    bool unsigned = 3;
}

message Expr {
    ExprType expr_type      = 1;    // uint32
    ColumnInfo column       = 2;    // left in optimization
    bytes value             = 3;    // maybe left or right, but right in optimization
    repeated Expr  child    = 4;    // alwayse 2 expr left and right, but LogicNot
}
